### PR TITLE
fix: hasher

### PR DESCRIPTION
**Pull request by SwenSchaeferjohann**

### CODE REVIEW

File: macros/light-sdk-macros/tests/hasher.rs [hunk 1]
```diff
@@ -28,51 +29,392 @@ pub struct MyNestedNonHashableStruct {
...
+                two: create_test_nested_struct(),
+            };
+
+            let manual_hash = Poseidon::hashv(&[
+                &test_nested_struct.one.hash::<Poseidon>().unwrap(),
+                &test_nested_struct.two.hash::<Poseidon>().unwrap(),
+            ])
+            .unwrap();
+
+            assert_eq!(test_nested_struct.hash::<Poseidon>().unwrap(), manual_hash);
```
Comment (ananas-block) [inline macros/light-sdk-macros/tests/hasher.rs:242]: can you add a hardcoded reference value for this and the tests that you changed?
Can you add a truncated value inside `MyNestedStruct`?

File: macros/light-sdk-macros/tests/hasher.rs [hunk 2]
```diff
@@ -28,51 +29,392 @@ pub struct MyNestedNonHashableStruct {
...
+            }
+
+            let nested = create_test_nested_struct();
+            let test_struct = NestedTruncate {
+                inner: nested,
+                data: "test".to_string(),
+            };
+
+            let hash = test_struct.hash::<Poseidon>().unwrap();
+            assert_eq!(hash.len(), 32);
```
Comment (ananas-block) [inline macros/light-sdk-macros/tests/hasher.rs:212]: imo we should add a manual comparison as well to make sure that the combination of nested and truncated really works.

File: macros/light-sdk-macros/tests/hasher.rs [hunk 3]
```diff
@@ -28,51 +29,392 @@ pub struct MyNestedNonHashableStruct {
...
-            b: Rc::new(RefCell::new(usize::MAX)),
-        },
-        f: Some(0),
-    };
-    assert_eq!(
-        my_account.hash::<Poseidon>().unwrap(),
-        [
-            32, 205, 141, 227, 236, 5, 28, 219, 24, 164, 215, 79, 151, 131, 162, 82, 224, 101, 171,
-            201, 4, 181, 26, 146, 6, 1, 95, 107, 239, 19, 233, 80
-        ]
```
Comment (ananas-block) [inline macros/light-sdk-macros/tests/hasher.rs:76]: pls add hardcoded references again

**Review by ananas-block — COMMENTED** [review]
Thanks looks pretty good already, added comments for additional tests.

**Comment by SwenSchaeferjohann** [comment]
do not merge yet

todo
* cover edge cases for option+unnested inner_struct / truncate+option

File: macros/light-sdk-macros/src/hasher.rs [hunk 4]
```diff
@@ -17,28 +17,84 @@ pub(crate) fn hasher(input: ItemStruct) -> Result<TokenStream> {
...
+                            result.push([0u8; 32].to_vec());
+                        }
+                    }
+                }
+            } else if is_option {
+                quote! {
+                    match &self.#field_name {
+                        Some(value) => {
+                            let bytes = value.to_le_bytes();
+                            // if bytes.len() == 32 {
```
Comment (vadorovsky) [inline macros/light-sdk-macros/src/hasher.rs:83]:
```suggestion
                            // if bytes.len() >= 32 {
```

And I think we should throw such an error :D

File: macros/light-sdk-macros/src/hasher.rs [hunk 5]
```diff
@@ -17,28 +17,84 @@ pub(crate) fn hasher(input: ItemStruct) -> Result<TokenStream> {
...
+
+
+            Ok(if nested {
+                if is_option {
+                    quote! {
+                        match &self.#field_name {
+                            Some(value) => {
+                                let mut nested_hash = ::light_hasher::DataHasher::hash::<::light_hasher::Poseidon>(value)
+                                    .expect("Failed to hash nested field");
+                                nested_hash[0] = 1; // Mark as Some
```
Comment (vadorovsky) [inline macros/light-sdk-macros/src/hasher.rs:49]: That's neat!

What I was thinking about before prepending `0` or `1` to the input of `hash` or `hash_to_bn254_field_size_be`, but actually this trick lets us to do that without additional allocations. I like that.

Comment (SwenSchaeferjohann) [reply macros/light-sdk-macros/src/hasher.rs]: spoke with @ananas-block, we figured the most robust solution is: for nested or truncated options: Some -> hash, None -> 0

Comment (vadorovsky) [reply macros/light-sdk-macros/src/hasher.rs]: That sounds great, in that case there is no chance of collision.

File: examples/name-service/programs/name-service/src/lib.rs [hunk 6]
```diff
@@ -76,6 +77,44 @@ pub struct NameRecord {
     #[truncate]
     pub name: String,
     pub rdata: RData,
+    #[nested]
+    pub nested: NestedData,
+}
+
+// Illustrates nested hashing feature.
+#[derive(LightHasher, Clone, Debug, AnchorSerialize, AnchorDeserialize)]
+pub struct NestedData {
```
Comment (ananas-block) [inline examples/name-service/programs/name-service/src/lib.rs:86]: integration tests are in `light-protocol/test-programs`. We should make a macro test program there.

File: test-programs/sdk-test-program/tests/test.rs [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+        &merkle_context,
+        &address_merkle_context,
+        &account_compression_authority,
+        &registered_program_pda,
+        &PROGRAM_ID_LIGHT_SYSTEM,
+    )
+    .await
+    .unwrap();
+
+    println!("done");
```
Comment (vadorovsky) [inline test-programs/sdk-test-program/tests/test.rs:96]: Leftover print?

File: test-programs/sdk-test-program/tests/test.rs [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+    let record = &compressed_account
+        .compressed_account
+        .data
+        .as_ref()
+        .unwrap()
+        .data;
+    let record = MyCompressedAccount::deserialize(&mut &record[..]).unwrap();
+    assert_eq!(record.nested.one, 1);
+
+    println!("done 2");
```
Comment (vadorovsky) [inline test-programs/sdk-test-program/tests/test.rs:111]: Leftover print?

File: test-programs/sdk-test-program/tests/test.rs [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+    // update_record(
```
Comment (vadorovsky) [inline test-programs/sdk-test-program/tests/test.rs:112]: I would remove this commented code - it's from a different example (DNS) and I think we can just add an update operation here later. I can take it as a next PR from my side.

Comment (SwenSchaeferjohann) [reply test-programs/sdk-test-program/tests/test.rs]: thanks - i added test coverage now for update_nested_data for completeness

File: macros/light-sdk-macros/tests/hasher.rs [hunk 10]
```diff
@@ -15,64 +16,1202 @@ pub struct MyAccount {
...
+                d: [0; 32],
+                e: MyNestedNonHashableStruct {
+                    a: PhantomData,
+                    b: Rc::new(RefCell::new(0)),
+                },
+                f: Some(0),
+            };
+
+            let hash = zero_account.hash::<Poseidon>().unwrap();
+            assert_eq!(hash.len(), 32);
```
Comment (ananas-block) [inline macros/light-sdk-macros/tests/hasher.rs:159]: this assert doesn't make sense to me since there is no case where the hash can be of a different length than 32 bytes.
```suggestion
              let manual_account_bytes: Vec<Vec<u8>> = vec![
                  vec![u8::from(zero_account.a)],
                  zero_account.b.to_le_bytes().to_vec(),
                  zero_account.c.hash::<Poseidon>().unwrap().to_vec(),
                  light_utils::hash_to_bn254_field_size_be(&zero_account.d)
                      .unwrap()
                      .0
                      .to_vec(),
                  zero_account.e.b.to_le_bytes().to_vec(),
                  {
                      let mut bytes = vec![1u8];
                      bytes.extend_from_slice(&zero_account.f.unwrap().to_le_bytes());
                      bytes
                  },
              ];
            let account_bytes: Vec<&[u8]> =
                manual_account_bytes.iter().map(|v| v.as_slice()).collect();
            let manual_account_hash = Poseidon::hashv(&account_bytes).unwrap();
            let hash = zero_account.hash::<Poseidon>().unwrap();
            assert_eq!(hash, manual_account_hash);
```

**Review by vadorovsky — APPROVED** [review]

File: macros/light-sdk-macros/tests/hasher.rs [hunk 11]
```diff
@@ -15,64 +16,1202 @@ pub struct MyAccount {
...
+        use super::*;
+
+        #[test]
+        fn test_byte_representation() {
+            let nested_struct = create_nested_struct();
+            let account = create_account(Some(42));
+
+            let manual_nested_bytes: Vec<Vec<u8>> = vec![
+                nested_struct.a.to_le_bytes().to_vec(),
+                nested_struct.b.to_le_bytes().to_vec(),
```
Comment (ananas-block) [inline macros/light-sdk-macros/tests/hasher.rs:118]: the poseidon hasher hashes everything be not le we should use be in the macros as well.

Comment (SwenSchaeferjohann) [reply macros/light-sdk-macros/tests/hasher.rs:117]: this will require changes across multiple crates. will change in a follow up pr + hold hasher release until then
